Applications/Spire/Source/Ui/ResponsiveLabel.cpp

```diff
@@ -0,0 +1,230 @@
+    property.visit(
+      [&] (const Font& font) {
+        stylist.evaluate(font, [&] (const auto& font) {
+          updated_font = font;
+        });
+      },
+      [&] (const FontSize& size) {
+        stylist.evaluate(size, [&] (auto size) {
+          updated_font_size = size;
+        });
+      });
+  }
+  if(updated_font != m_text_box_font ||
+      updated_font_size != m_text_box_font.pixelSize()) {
+    m_text_box_font = updated_font;
+    m_text_box_font.setPixelSize(updated_font_size);
+    auto current = m_current_mapped_index;
+    reset_mapped_labels();
+    set_current(current);
+  }
+}
+
+void ResponsiveLabel::update_display_text() {
+  if(m_mapped_labels.empty()) {
+    set_current(none);
+    return;
+  } else if(m_mapped_labels.back().m_pixel_width <= width()) {
+    set_current(m_mapped_labels.size() - 1);
+    return;
+  }
+  auto mapped_label = [&] {
+    return std::lower_bound(m_mapped_labels.begin(), m_mapped_labels.end(),
+      width() + 1, [=] (const auto& mapped_label, auto width) {
+        return mapped_label.m_pixel_width < width;
+      });
+  }();
+  if(mapped_label == m_mapped_labels.begin()) {
+    set_current(0);
+    return;
+  }
+  set_current(std::distance(m_mapped_labels.begin(), std::prev(mapped_label)));
+}
+
+void ResponsiveLabel::on_label_added(int index) {
+  m_mapped_labels.push_back({index, get_pixel_width(m_labels->get(index))});
+  if(m_mapped_labels.size() > 1 && m_mapped_labels.back().m_pixel_width <
+      m_mapped_labels.at(m_mapped_labels.size() - 2).m_pixel_width) {
+    sort_mapped_labels();
+  }
+  set_current(m_mapped_labels.size() - 1);
+}
+
+void ResponsiveLabel::on_label_removed(int index) {
+  for(auto i = m_mapped_labels.begin(); i != m_mapped_labels.end(); ++i) {
+    if(i->m_list_index == index) {
+      i = m_mapped_labels.erase(i);
+      if(i == m_mapped_labels.end()) {
+        break;
+      }
+    }
+    if(i->m_list_index > index) {
+      --(i->m_list_index);
+    }
+  }
+  auto current_index = [&] () -> optional<int> {
+    if(m_mapped_labels.empty()) {
+      return none;
+    }
+    return m_mapped_labels.size() - 1;
+  }();
+  set_current(current_index);
+}
+
+void ResponsiveLabel::on_label_updated(int index) {
+  auto mapped_label =
+    std::find_if(m_mapped_labels.begin(), m_mapped_labels.end(),
+      [&] (const auto& mapped_label) {
+        return index == mapped_label.m_list_index;
+      });
+  mapped_label->m_pixel_width = get_pixel_width(m_labels->get(index));
+  if(m_mapped_labels.size() > 1) {
+    auto lower = [&] {
+      if(mapped_label == m_mapped_labels.begin()) {
+        return 0;
+      }
+      return std::prev(mapped_label)->m_pixel_width;
+    }();
+    auto upper = [&] {
+      if(std::next(mapped_label) == m_mapped_labels.end()) {
+        return mapped_label->m_pixel_width;
+      }
+      return std::next(mapped_label)->m_pixel_width;
+    }();
+    if(mapped_label->m_pixel_width <= lower ||
```

Comment:
When `m_pixel_width == lower` or `m_pixel_width == upper`, do we need to sort labels?

Reply:
I believe we do because of the case where two equal length labels need to be sorted by list index.

The case where they would be equal would be something like a list: [CCC, AA, BB, DD]. Assuming each single character has the same pixel width, the mapped labels would be sorted as [AA, BB, DD, CCC]. If CCC was then updated to be 'CC', it would become equal in pixel width to the previous mapped label, DD, and the mapped labels would then need to be sorted to [CC, AA, BB, DD].